### PR TITLE
qa: Enable flake8 tox and fix failures

### PR DESCRIPTION
**Pull request by toabctl**
There were a couple of problems found by flake8 in the qa/
directory (most of them fixed now). Enabling flake8 during the usual
check runs hopefully avoids adding new issues in the future.

Signed-off-by: Thomas Bechtold <tbechtold@suse.com>

### CODE REVIEW

File: qa/tasks/ceph2.py [hunk 1]
```diff
@@ -724,7 +709,9 @@ def restart(ctx, config):
             healthy(ctx=ctx, config=dict(cluster=cluster))
     if config.get('wait-for-osds-up', False):
         for cluster in clusters:
-            wait_for_osds_up(ctx=ctx, config=dict(cluster=cluster))
+            # FIXME this method is not defined!!!
+            # wait_for_osds_up(ctx=ctx, config=dict(cluster=cluster))
```
Comment (sebastian-philipp) [inline qa/tasks/ceph2.py]: @sage: *`wait_for_osds_up`: FIXME this method is not defined!!!*

Comment (liewegas) [reply qa/tasks/ceph2.py]: should be
```
            ctx.managers[cluster].wait_for_all_osds_up()
```

File: qa/tasks/cephfs/test_cephfs_shell.py [hunk 2]
```diff
@@ -611,8 +609,8 @@ def test_du_with_path_in_args(self):
             path_prefix='')
 
         args = ['du', '/']
-        for path in path_to_files:
-            args.append(path)
+        for p in path_to_files:
+            args.append(p)
```
Comment (kshtsk) [inline qa/tasks/cephfs/test_cephfs_shell.py]: why this changes is needed? even though, why not just:
```args.append(path_to_files)```

File: qa/tasks/cephfs/test_exports.py [hunk 3]
```diff
@@ -132,8 +132,8 @@ def test_export_pin_getfattr(self):
             p = self.mount_a.client_remote.run(args=['uname', '-r'], stdout=StringIO(), wait=True)
```
Comment (kshtsk) [inline qa/tasks/cephfs/test_exports.py]: can this line be changed to:
```
client_uname = self.mount_a.client_remote.sh('unman -r', wait=True)
```
and later:
```
if client_uname < "5" and not(dir_pin):
. . .
```

File: qa/tasks/cephfs/test_exports.py [hunk 4]
```diff
@@ -150,7 +150,7 @@ def test_session_race(self):
         status = self.fs.wait_for_daemons()
 
         rank1 = self.fs.get_rank(rank=1, status=status)
-        name1 = 'mds.'+rank1['name']
+        _ = 'mds.'+rank1['name']
```
Comment (kshtsk) [inline qa/tasks/cephfs/test_exports.py]: if it is not needed maybe delete? or comment out?

File: qa/tasks/cephfs/test_volume_client.py [hunk 5]
```diff
@@ -1078,7 +1077,7 @@ def test_volume_without_namespace_isolation(self):
         volume_prefix = "/myprefix"
         group_id = "grpid"
         volume_id = "volid"
-        mount_path = self._volume_client_python(vc_mount, dedent("""
+        _ = self._volume_client_python(vc_mount, dedent("""
```
Comment (kshtsk) [inline qa/tasks/cephfs/test_volume_client.py]: if the right part needed, then maybe just remove left part of?

File: qa/tasks/cephfs/xfstests_dev.py [hunk 6]
```diff
@@ -118,7 +119,7 @@ def install_deps(self):
         else:
             raise RuntimeError('expected a yum based or a apt based system')
 
-        proc = self.mount_a.client_remote.run(args=args, omit_sudo=False)
+        _ = self.mount_a.client_remote.run(args=args, omit_sudo=False)
```
Comment (kshtsk) [inline qa/tasks/cephfs/xfstests_dev.py]: just remove left side

File: qa/tasks/cram.py [hunk 7]
```diff
@@ -124,7 +126,7 @@ def _run_tests(ctx, role):
     :param ctx: Context
     :param role: Roles
     """
-    assert isinstance(role, basestring)
+    assert isinstance(role, six.string_types)
```
Comment (kshtsk) [inline qa/tasks/cram.py]: I hope we can do `isinstance(role, str)` I would prefer to use six as less as possible.

File: qa/tasks/manypools.py [hunk 8]
```diff
@@ -46,28 +46,28 @@ def task(ctx, config):
...
+                    '--name', role_,
+                    'osd', 'pool', 'create', 'pool{num}'.format(num=poolnum), '8',
+                    run.Raw('&&'),
+                    'rados',
+                    '--name', role_,
+                    '--pool', 'pool{num}'.format(num=poolnum),
+                    'bench', '0', 'write', '-t', '16', '--block-size', '1'
+                ],
+                wait = False
+            )
```
Comment (kshtsk) [inline qa/tasks/manypools.py]: As far as you touch each of the line, can this be replaced with human readable string, like?
```
proc = remote.run(args=
   "ceph --name {role} osd pool create pool{num} 8 && "
   "rados --name {role} --pool pool{num} bench 0 "
    "write -t 16 --block-size 1".format(role=role, num=poolnum), wait=False)
```

File: qa/tasks/mgr/dashboard/test_cephfs.py [hunk 9]
```diff
@@ -62,7 +62,7 @@ def test_cephfs_clients(self):
 
     def test_cephfs_evict_client_does_not_exist(self):
         fs_id = self.get_fs_id()
-        data = self._delete("/api/cephfs/{}/client/1234".format(fs_id))
+        _ = self._delete("/api/cephfs/{}/client/1234".format(fs_id))
```
Comment (kshtsk) [inline qa/tasks/mgr/dashboard/test_cephfs.py]: why do we need '_ ='?

File: qa/tasks/mgr/dashboard/test_rbd.py [hunk 10]
```diff
@@ -764,7 +764,7 @@ def test_move_image_to_trash(self):
         id = self.create_image_in_trash('rbd', 'test_rbd')
         self.assertStatus(200)
 
-        img = self.get_image('rbd', None, 'test_rbd')
+        _ = self.get_image('rbd', None, 'test_rbd')
```
Comment (kshtsk) [inline qa/tasks/mgr/dashboard/test_rbd.py]: remove '_ ='

File: qa/tasks/vstart_runner.py [hunk 11]
```diff
@@ -1215,7 +1214,7 @@ def scan_tests(modules):
     max_required_mgr = 0
     require_memstore = False
 
-    for suite, case in enumerate_methods(overall_suite):
+    for suite_, case in enumerate_methods(overall_suite):
```
Comment (kshtsk) [inline qa/tasks/vstart_runner.py]: ?
```for case in enumerate_methods(overall_suite).values()```

File: qa/tasks/cephfs/test_cephfs_shell.py (hunk 2 above)
Comment (toabctl) [reply qa/tasks/cephfs/test_cephfs_shell.py]: to not overlay `path` which is already defined at the top with `from os import path`

File: qa/tasks/cephfs/test_exports.py (hunk 4 above)
Comment (toabctl) [reply qa/tasks/cephfs/test_exports.py]: Done

File: qa/tasks/cephfs/test_volume_client.py (hunk 5 above)
Comment (toabctl) [reply qa/tasks/cephfs/test_volume_client.py]: Done

File: qa/tasks/cephfs/xfstests_dev.py (hunk 6 above)
Comment (toabctl) [reply qa/tasks/cephfs/xfstests_dev.py]: done

File: qa/tasks/ceph2.py (hunk 1 above)
Comment (toabctl) [reply qa/tasks/ceph2.py]: done

File: qa/tasks/cram.py (hunk 7 above)
Comment (toabctl) [reply qa/tasks/cram.py]: Why? `six` is exactly there for theses cases. Imo there is no reason to avoid it (given that we already have it in the teuthology install_requires). Plus it's then easy to find the py2/py3 compat code if we are going to remove the python2 part.

File: qa/tasks/manypools.py (hunk 8 above)
Comment (toabctl) [reply qa/tasks/manypools.py]: I want to avoid that. My goal is to make flake8 happy for python2 and python3. I want to change as less as possible.

File: qa/tasks/mgr/dashboard/test_rbd.py (hunk 10 above)
Comment (toabctl) [reply qa/tasks/mgr/dashboard/test_rbd.py]: done

File: qa/tasks/vstart_runner.py (hunk 11 above)
Comment (toabctl) [reply qa/tasks/vstart_runner.py]: suite is already used by `from unittest import suite` so renaming the variable will solve that.

File: qa/tasks/mgr/dashboard/test_cephfs.py (hunk 9 above)
Comment (toabctl) [reply qa/tasks/mgr/dashboard/test_cephfs.py]: done

File: qa/tasks/cephfs/test_exports.py (hunk 3 above)
Comment (toabctl) [reply qa/tasks/cephfs/test_exports.py]: I would like to avoid that in this PR. There are plenty of `StringIO` imports left that are not py3 compatible. Can we tackle these in a different PR?

File: qa/tasks/cephfs/filesystem.py [hunk 12]
```diff
@@ -778,7 +778,9 @@ def get_daemon_names(self, state=None, status=None):
         """
         mdsmap = self.get_mds_map(status)
         result = []
-        for mds_status in sorted(mdsmap['info'].values(), lambda a, b: cmp(a['rank'], b['rank'])):
+        for mds_status in sorted(
+                mdsmap['info'].values(),
+                lambda a, b: (a['rank'] > b['rank']) - (a['rank'] < b['rank'])):
```
Comment (kshtsk) [inline qa/tasks/cephfs/filesystem.py]: Why not just use key?
```
for mds_status in sorted(mdsmap['info'].values(), key=lambda _: _['rank']):
```

File: qa/tasks/cephfs_test_runner.py [hunk 13]
```diff
@@ -133,7 +133,9 @@ def task(ctx, config):
 
     # Mount objects, sorted by ID
     if hasattr(ctx, 'mounts'):
-        mounts = [v for k, v in sorted(ctx.mounts.items(), lambda a, b: cmp(a[0], b[0]))]
+        mounts = [v for k, v in sorted(
+            ctx.mounts.items(),
+            lambda a, b: (a[0] > b[0]) - (a[0] < b[0]))]
```
Comment (kshtsk) [inline qa/tasks/cephfs_test_runner.py]: Use key for py2/py3 compatibility:
```
mounts = [v for k, v in sorted(ctx.mount.items(), key=lambda _: _[0])]
```

File: qa/tasks/cephfs/filesystem.py (hunk 12 above)
Comment (toabctl) [reply qa/tasks/cephfs/filesystem.py]: I just followed the python docs how to replace `cmp`. See https://docs.python.org/3.0/whatsnew/3.0.html#ordering-comparisons

File: qa/tasks/cephfs_test_runner.py (hunk 13 above)
Comment (toabctl) [reply qa/tasks/cephfs_test_runner.py]: I would prefer todo that in another PR. I'm sure there are plenty of problems with python3. This is just a first step to catch the errors that flake8 (run on python3) can catch for us.

File: qa/tasks/cephfs/filesystem.py (hunk 12 above)
Comment (kshtsk) [reply qa/tasks/cephfs/filesystem.py]: @toabctl I understand, and one of the item from this guide says: **Use the key argument instead.**, see the cite:
```
builtin.sorted() and list.sort() no longer accept the cmp argument 
providing a comparison function. Use the key argument instead. 
N.B. the key and reverse arguments are now “keyword-only”.
```

Comment (toabctl) [reply qa/tasks/cephfs/filesystem.py]: Another cite:
```
If you really need the cmp() functionality, you could use the expression (a > b) - (a < b) as the equivalent for cmp(a, b).
```
And given that I want to change as less as possible I decided to go that way.

File: qa/tasks/cephfs_test_runner.py (hunk 13 above)
Comment (kshtsk) [reply qa/tasks/cephfs_test_runner.py]: This is very cheap change. It is much easier to change it here now, just with git commit --amend, than rerun the tests later.

Comment (toabctl) [reply qa/tasks/cephfs_test_runner.py]: It's unrelated to the commit messages and unrelated to this PR. This is just a single py2/py3 compat change and there are maybe 100 more needed. The question for me is: How can we

- fix all theses py2/py3 compat issues and
- make sure that no new regressions for python3 are introduced

If you really want me to do this change, I can do it. But I'm not in favor of it.

**Comment by toabctl** [comment]
jenkins test make check

File: qa/tasks/cephfs_test_runner.py (hunk 13 above)
Comment (kshtsk) [reply qa/tasks/cephfs_test_runner.py]: from my point it is directly related to this commit message and PR, because it removes cmp usage the same way you did it

**Comment by toabctl** [comment]
I dropped the python3 flake8 run commit for now (there are problems with the venvs).

File: qa/tasks/cephfs/test_cephfs_shell.py (hunk 2 above)
Comment (kshtsk) [reply qa/tasks/cephfs/test_cephfs_shell.py]: looks reasonable, anyways, could you please use one-liner args.append instead of for loop?

File: qa/tasks/cephfs/test_scrub_checks.py [hunk 14]
```diff
@@ -307,7 +307,7 @@ def tell_command(self, mds_rank, command, validator):
 
         success, errstring = validator(jout, 0)
         if not success:
-            raise AsokCommandFailedError(command, rout, jout, errstring)
+            raise AsokCommandFailedError(command, success, jout, errstring)
```
Comment (tchaikov) [inline qa/tasks/cephfs/test_scrub_checks.py]: should pass "0" instead of "success" as the return code. as success here is "False" instead of a number.

**Comment by toabctl** [comment]
@liewegas imo ready for merging now.

File: qa/tasks/cephfs/test_scrub_checks.py [hunk 15]
```diff
@@ -43,7 +43,7 @@ def test_scrub_abort(self):
         log.info("client_path: {0}".format(client_path))
 
         log.info("Cloning repo into place")
-        repo_path = TestScrubChecks.clone_repo(self.mount_a, client_path)
+        _ = TestScrubChecks.clone_repo(self.mount_a, client_path)
```
Comment (tchaikov) [inline qa/tasks/cephfs/test_scrub_checks.py]: just out of curiosity, why not just ignore the return value?

File: qa/tasks/cephfs/test_scrub_checks.py (hunk 14 above)
Comment (toabctl) [reply qa/tasks/cephfs/test_scrub_checks.py]: done

File: qa/tasks/cephfs/test_scrub_checks.py (hunk 15 above)
Comment (toabctl) [reply qa/tasks/cephfs/test_scrub_checks.py]: Done

**Comment by sebastian-philipp** [comment]
I guess we need an exhaustive QA run for this PR?

File: qa/tasks/cephfs/test_scrub_checks.py (hunk 15 above)
Comment (tchaikov) [reply qa/tasks/cephfs/test_scrub_checks.py]: there more occurrences of this pattern in this changeset. but i think it's fine.

File: qa/tasks/cephfs/filesystem.py (hunk 12 above)
Comment (kshtsk) [reply qa/tasks/cephfs/filesystem.py]: @toabctl yes, but the cmp is not required here, IMHO.

Comment (kshtsk) [reply qa/tasks/cephfs/filesystem.py]: I suggest to do things right way from with on PR instead of having a bunch of overlapping commits.

File: qa/tasks/cephfs_test_runner.py (hunk 13 above)
Comment (tchaikov) [reply qa/tasks/cephfs_test_runner.py]: IMHO, this is not in the scope of pep8 compliance. probably we can leave the py3 compatible change to another commit or PR?

Comment (kshtsk) [reply qa/tasks/cephfs_test_runner.py]: This is exactly about py3 compatibility, since this change required for python3 flake8, if it is not a scope py3 than let's run py27 flake8 and it will not complain against this line and this change is not required then :-)

Comment (tchaikov) [reply qa/tasks/cephfs_test_runner.py]: i see. `run_tox.sh` uses both python2 and python3 envs when running the tox test.

Comment (toabctl) [reply qa/tasks/cephfs_test_runner.py]: FYI - I already removed the 2nd commit which included this change. So this is no longer relevant. But my plan is to propose the flake8-python3 run (including the fixes) in another PR and then this gets relevant again.

**Comment by toabctl** [comment]
rebased...

File: qa/tasks/vstart_runner.py (hunk 11 above)
Comment (kshtsk) [reply qa/tasks/vstart_runner.py]: maybe better to use unittest.suite in this case?

File: qa/tasks/cephfs/filesystem.py (hunk 12 above)
Comment (toabctl) [reply qa/tasks/cephfs/filesystem.py]: No longer part of this PR so marking this as resolved

File: qa/tasks/vstart_runner.py (hunk 11 above)
Comment (toabctl) [reply qa/tasks/vstart_runner.py]: what's wrong with `suite_`?

File: qa/tasks/cram.py (hunk 7 above)
Comment (toabctl) [reply qa/tasks/cram.py]: No longer part of this PR so marking this as resolved

File: qa/tasks/cephfs/test_exports.py (hunk 3 above)
Comment (toabctl) [reply qa/tasks/cephfs/test_exports.py]: No longer part of this PR so marking this as resolved

File: qa/tasks/cephfs/test_failover.py [hunk 16]
```diff
@@ -360,7 +358,7 @@ class TestStandbyReplay(CephFSTestCase):
 
     def _confirm_no_replay(self):
         status = self.fs.status()
-        standby_count = len(list(status.get_standbys()))
+        _ = len(list(status.get_standbys()))
```
Comment (kshtsk) [inline qa/tasks/cephfs/test_failover.py]: Why not just remove this line if it is not used?
